src/main/java/org/folio/service/budget/RolloverBudgetExpenseClassTotalsService.java

```diff
@@ -0,0 +1,113 @@
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+import static java.util.stream.Collectors.toMap;
+import static java.util.stream.Collectors.groupingBy;
+
+public class RolloverBudgetExpenseClassTotalsService {
+
+  private final BudgetExpenseClassService budgetExpenseClassService;
+  private final TemporaryTransactionService temporaryTransactionService;
+
+  public RolloverBudgetExpenseClassTotalsService(BudgetExpenseClassService budgetExpenseClassService, TemporaryTransactionService temporaryTransactionService) {
+    this.budgetExpenseClassService = budgetExpenseClassService;
+    this.temporaryTransactionService = temporaryTransactionService;
+  }
+
+  public Future<LedgerFiscalYearRolloverBudget> getBudgetWithUpdatedExpenseClassTotals(LedgerFiscalYearRolloverBudget budget, DBClient dbClient) {
+    return budgetExpenseClassService.getExpenseClassesByTemporaryBudgetId(budget.getId(), dbClient)
+      .compose(expenseClasses -> temporaryTransactionService.getTransactions(budget, dbClient)
+        .map(transactions -> buildBudgetExpenseClassesTotals(expenseClasses, transactions, budget)))
+      .compose(budgetExpenseClassTotals -> budgetExpenseClassService.getTempBudgetExpenseClasses(budget.getId(), dbClient)
+        .map(budgetExpenseClasses -> updateExpenseClassStatus(budgetExpenseClassTotals, budgetExpenseClasses)))
+      .map(budget::withExpenseClassDetails);
+  }
+
+  private List<BudgetExpenseClassTotal> buildBudgetExpenseClassesTotals(List<ExpenseClass> expenseClasses, List<Transaction> transactions, LedgerFiscalYearRolloverBudget budget) {
+
+    double totalExpended = getBudgetTotalExpended(budget);
+
+    Map<String, List<Transaction>> groupedByExpenseClassId = transactions.stream()
+      .filter(transaction -> Objects.nonNull(transaction.getExpenseClassId()))
+      .collect(groupingBy(Transaction::getExpenseClassId));
+
+    Map<ExpenseClass, List<Transaction>> groupedByExpenseClass = expenseClasses.stream()
+      .collect(toMap(Function.identity(), expenseClass -> groupedByExpenseClassId.getOrDefault(expenseClass.getId(), Collections.emptyList())));
+
+    return buildBudgetExpenseClassesTotals(groupedByExpenseClass, totalExpended);
+  }
+
+  private double getBudgetTotalExpended(LedgerFiscalYearRolloverBudget budget) {
+    BigDecimal totalExpended = BigDecimal.valueOf(budget.getExpenditures()).add(BigDecimal.valueOf(budget.getOverExpended()));
+    return totalExpended.doubleValue();
+  }
+
+  private List<BudgetExpenseClassTotal> buildBudgetExpenseClassesTotals(Map<ExpenseClass, List<Transaction>> groupedByExpenseClass, double totalExpended) {
+    return groupedByExpenseClass.entrySet().stream()
+      .map(entry -> buildBudgetExpenseClassTotals(entry.getKey(), entry.getValue(), totalExpended))
+      .collect(Collectors.toList());
+  }
+
+  private BudgetExpenseClassTotal buildBudgetExpenseClassTotals(ExpenseClass expenseClass, List<Transaction> transactions, double totalExpended) {
```

Comment:
comments describing each block in this method will go a long way to help readability.

Reply:
It is hard to describe when I'm not an author of the original logic: https://github.com/folio-org/mod-finance/blob/master/src/main/java/org/folio/services/budget/BudgetExpenseClassTotalsService.java

Reply:
I plan to create a task to move the logic from mod-finance to mod-finance-storage to avoid duplication.

Reply:
Sounds good. Since you can't describe it as well, it shows there is a problem here. Please create the task.